src/Security/InteractiveSecurityHelper.php

```diff
@@ -140,22 +143,57 @@ public function guessPasswordField(SymfonyStyle $io, string $userClass): string
         );
     }
 
-    public function getAuthenticatorClasses(array $firewallData): array
+    /**
+     * @param array<string, array<string, mixed>> $firewalls Config data from security.firewalls
+     *
+     * @return Authenticator[]
+     */
+    public function getAuthenticatorsFromConfig(array $firewalls): array
     {
-        if (isset($firewallData['guard'])) {
-            return array_filter($firewallData['guard']['authenticators'] ?? [], static fn ($authenticator) => class_exists($authenticator));
-        }
+        $authenticators = [];
 
-        if (isset($firewallData['custom_authenticator'])) {
-            $authenticators = $firewallData['custom_authenticator'];
-            if (\is_string($authenticators)) {
-                $authenticators = [$authenticators];
+        // Iterate of each firewall that exists e.g. security.firewalls.main
+        foreach ($firewalls as $firewallName => $firewallConfig) {
+            if (!\is_array($firewallConfig)) {
+                continue;
             }
 
-            return array_filter($authenticators, static fn ($authenticator) => class_exists($authenticator));
+            foreach ($firewallConfig as $potentialAuthenticator => $configData) {
+                if (null === ($authenticator = AuthenticatorType::tryFrom($potentialAuthenticator))) {
+                    // This entry is probably security.firewalls.main.lazy or security.firewalls.main.providers
+                    continue;
+                }
+
+                if (AuthenticatorType::CUSTOM !== $authenticator) {
+                    // We found a "built in" authenticator - "form_login", "json_login", etc...
+                    $authenticators[] = new Authenticator($authenticator, $firewallName);
+
+                    continue;
+                }
+
+                // custom_authenticators can be set as a string or an array in security.yaml
+                if (\is_string($configData)) {
+                    $configData = [$configData];
+                }
+
+                foreach ($configData as $customAuthenticatorClass) {
+                    //                    if (!class_exists($customAuthenticatorClass)) {
```

Comment:
I'm not sure if we _should_ check if the authenticator exists and that it implements `AuthenticatorInterface::class`. Doing so would be a bonus? but it would also require refactoring the test suite to implement a concrete `CustomAuthenticatorFixture::class` Thoughts?

Reply:
I'm actually a bit lost in this method. We're looping over the keys under a firewall, right? And the first if statement with `tryFrom` will exit unless the key matches the specific set we have in `AuthenticatorType`, right? So then, `$configData` is represents the values beneath those keys. So if I have:

```
firewalls:
    main:
        form_login:
            path: /login
```

Won't `$configData` at this point be `['path' => '/login']`? Or am I totally misreading the method?